components/renku_data_services/session/db.py

```diff
@@ -654,24 +623,73 @@ async def __update_launcher_environment(
+                launcher.environment.command = update.command
+                launcher.environment.args = update.args
+                launcher.environment.environment_image_source = models.EnvironmentImageSource.image
+                launcher.environment.build_parameters_id = None
+
+                # NOTE: Delete the build parameters since they are not used by any other environment
+                await session.delete(build_parameters)
+
+                await session.flush()
+            case models.UnsavedBuildParameters() as new_custom_built_environment, models.EnvironmentKind.GLOBAL:
                 # Global environment replaced by a custom one which will be built
                 new_env = self.__insert_build_parameters_environment(
-                    user, session, launcher, new_custom_build_parameters_environment
+                    user, session, launcher, new_custom_built_environment
                 )
+                launcher.environment_id = new_env.id
                 launcher.environment = new_env
+                await session.flush()
+            case models.UnsavedBuildParameters() as new_custom_built_environment, models.EnvironmentKind.CUSTOM:
+                # Custom environment with image is replaced by a custom environment with build
+                build_parameters_orm = schemas.BuildParametersORM(
+                    builder_variant=new_custom_built_environment.builder_variant,
+                    frontend_variant=new_custom_built_environment.frontend_variant,
+                    repository=new_custom_built_environment.repository,
+                )
+                session.add(build_parameters_orm)
+
+                launcher.environment.container_image = (
+                    "image:unknown-at-the-moment"  # TODO: This should come from the build
+                )
+                launcher.environment.default_url = "/lab"  # TODO: This should come from the build
+                launcher.environment.port = 8888  # TODO: This should come from the build
+                launcher.environment.working_directory = None  # TODO: This should come from the build
+                launcher.environment.mount_directory = None  # TODO: This should come from the build
+                launcher.environment.uid = 1000  # TODO: This should come from the build
+                launcher.environment.gid = 1000  # TODO: This should come from the build
+                launcher.environment.environment_kind = models.EnvironmentKind.CUSTOM
+                launcher.environment.command = None  # TODO: This should come from the build
+                launcher.environment.args = None  # TODO: This should come from the build
+                launcher.environment.environment_image_source = models.EnvironmentImageSource.build
+                launcher.environment.build_parameters_id = build_parameters_orm.id
+                launcher.environment.build_parameters = build_parameters_orm
```

Comment:
I know this is not really related to this PR(it was like this before) but I think we should handle this a bit better. Is there an issue for these TODOs? I think we'd either want some polymorphism in the db with an `UnbuiltEnvironment` type or to have a field in environment like `environment.state` that can be something like `not_built` and to allow the fields to be None (with dedicated types for built/unbuilt in code that validate that it's not None except for in this case)

Reply:
We'll do some work to store the build result and update the environment once a build finishes (there's a scope's task for that in the build note). I'll create an issue for it and add your comment there.